java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java

```diff
@@ -729,6 +745,146 @@ private void dumpBuffers(final StringBuilder sb, final Set<BufferLedger> ledgerS
     }
   }
 
+  /**
+   * Returns a default {@link Config} instance.
+   *
+   * @see ConfigBuilder
+   */
+  public static Config defaultConfig() {
+    return DEFAULT_CONFIG;
+
+  }
+
+  /**
+   * Returns a builder class for configuring BaseAllocator's options.
+   */
+  public static ConfigBuilder configBuilder() {
```

Comment:
Can you add [Immutables]( http://immutables.github.io/) to the project and use that instead. Should allow us to avoid creation/management of builders. I've started using it in another project and found it quite helpful at avoiding lots of scaffolding code.

Reply:
> Let's add a new method which is
> T unwrap(Class clazz)

Done.

> Can you add Immutables to the project and use that instead.

Thanks for suggesting Immutables, I have tried and it works pretty well. The only thing I am worried about is it slightly increases the development complexity. But I know a lot of projects already deal with the conditions in similar way so I have no objections on introducing it. 
Done.